Add alexaSmartHome event

horike37 opened the pull request:
<!--
1. Please check out and follow our Contributing Guidelines: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md
2. Do not remove any section of the template. If something is not applicable leave it empty but leave it in the PR
3. Please follow the template, otherwise we'll have to ask you to update it and it will take longer until your PR is merged
-->

## What did you implement:
Added `alexaSmartHome` event as a new one.

Closes #3955

<!--
Briefly describe the feature if no issue exists for this PR
-->

## How did you implement it:
I have created `alexaSmartHome` directory under the `events` one and implemented it.
https://github.com/serverless/serverless/tree/master/lib/plugins/aws/package/compile/events
<!--
If this is a nontrivial change please briefly describe your implementation so its easy for us to understand and review your code.
-->

## How can we verify it:
`amzn1.ask.skill.xx-xx-xx-xx` is an application ID for Alexa Smart Home Skill.
First, please sign up [Amazon Developer  Console](https://developer.amazon.com/) and get your application ID.
Then, write up a serverless.yml as follow, and run `sls deploy`

```yml
functions:
  mySkill:
    handler: mySkill.handler
    events:
      - alexaSmartHome: amzn1.ask.skill.xx-xx-xx-xx
```

or 

```yml
functions:
  mySkill:
    handler: mySkill.handler
    events:
      - alexaSmartHome: 
           appId: amzn1.ask.skill.xx-xx-xx-xx
           enabled: <true or false>
```
<!--
Add any applicable config, commands, screenshots or other resources
to make it easy for us to verify this works. The easier you make it for us
to review a PR, the faster we can review and merge it.

Examples:
* serverless.yml - Fully functioning to easily deploy changes
* Screenshots - Showing the difference between your output and the master
* Cloud Configuration - List cloud resources and show that the correct configuration is in place (e.g. AWS CLI commands)
* Other - Anything else that comes to mind to help us evaluate
-->

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

pmuens commented:
Great! Thanks for PRing this @horike37 👍 🎉

We'll review this ASAP 💪

pmuens commented:
Good catch @jvsinclair 👍

Yes, I agree. Adding a section in the docs should be sufficient here (and will avoid confusion).

@horike37 let us know if you need any help with that!

pmuens commented on docs/providers/aws/events/alexa-smart-home.md:
```diff
@@ -0,0 +1,43 @@
+## Event definition
+
+This will enable your Lambda function to be called by an Alexa Smart Home Skill.
+`amzn1.ask.skill.xx-xx-xx-xx` is an application ID for Alexa Smart Home. You need to sign up [Amazon Developer  Console](https://developer.amazon.com/) and get your application ID.
```
`Amazon Developer  Console` --> `Amazon Developer Console` (one space can be removed 😬).

pmuens commented on docs/providers/aws/guide/resources.md:
```diff
@@ -70,7 +70,7 @@ We're also using the term `normalizedName` or similar terms in this guide. This
 |Lambda::Version        | {normalizedFunctionName}LambdaVersion{sha256}                           | HelloLambdaVersionr3pgoTvv1xT4E4NiCL6JG02fl6vIyi7OS1aW0FwAI |
 |Logs::LogGroup         | {normalizedFunctionName}LogGroup                        | HelloLogGroup                 |
-|Lambda::Permission     | <ul><li>**Schedule**: {normalizedFunctionName}LambdaPermissionEventsRuleSchedule{index}</li><li>**CloudWatch Event**: {normalizedFunctionName}LambdaPermissionEventsRuleCloudWatchEvent{index}</li><li>**CloudWatch Log**: {normalizedFunctionName}LambdaPermissionLogsSubscriptionFilterCloudWatchLog{index}</li><li>**IoT**: {normalizedFunctionName}LambdaPermissionIotTopicRule{i ...
+|Lambda::Permission     | <ul><li>**Schedule**: {normalizedFunctionName}LambdaPermissionEventsRuleSchedule{index}</li><li>**CloudWatch Event**: {normalizedFunctionName}LambdaPermissionEventsRuleCloudWatchEvent{index}</li><li>**CloudWatch Log**: {normalizedFunctionName}LambdaPermissionLogsSubscriptionFilterCloudWatchLog{index}</li><li>**IoT**: {normalizedFunctionName}LambdaPermissionIotTopicRule{i ...
```
Could you add an example for the `Alexa Smart Home` resource name here (in the table all the examples are in the right column).

pmuens commented on lib/plugins/aws/package/compile/events/alexaSmartHome/index.js:
```diff
@@ -0,0 +1,87 @@
+            };
+
+            _.merge(this.serverless.service.provider.compiledCloudFormationTemplate.Resources,
+              permissionCloudForamtionResource);
```
`permissionCloudForamtionResource` --> `permissionCloudFormationResource`

pmuens commented on lib/plugins/aws/package/compile/events/alexaSmartHome/index.js:
```diff
@@ -0,0 +1,87 @@
+              .getLambdaAlexaSmartHomePermissionLogicalId(functionName,
+                alexaSmartHomeNumberInFunction);
+
+            const permissionCloudForamtionResource = {
```
`permissionCloudForamtionResource` --> `permissionCloudFormationResource`

pmuens commented on lib/plugins/aws/package/compile/events/alexaSmartHome/index.test.js:
```diff
@@ -0,0 +1,221 @@
+const AwsCompileAlexaSmartHomeEvents = require('./index');
+const Serverless = require('../../../../../../Serverless');
+
+describe('awsCompileAlexaSmartHomeEvents', () => {
```
`AwsCompileAlexaSmartHomeEvents` (the plugins class name)

horike37 commented:
@pmuens 
fixed what you pointed out and added the example stuff to `Kotlin ` template :smile:
Thank you for giving us the notice @jvsinclair. Added that to the doc :muscle:

jvsinclair commented:
@pmuens are you ok with his changes? I would be great to get this merged.

pmuens commented:
Thanks for pinging @jvsinclair 👍

I'll do a final review the upcoming days!

pmuens commented on docs/providers/aws/guide/resources.md:
```diff
@@ -70,7 +70,7 @@ We're also using the term `normalizedName` or similar terms in this guide. This
 |Lambda::Version        | {normalizedFunctionName}LambdaVersion{sha256}                           | HelloLambdaVersionr3pgoTvv1xT4E4NiCL6JG02fl6vIyi7OS1aW0FwAI |
 |Logs::LogGroup         | {normalizedFunctionName}LogGroup                        | HelloLogGroup                 |
-|Lambda::Permission     | <ul><li>**Schedule**: {normalizedFunctionName}LambdaPermissionEventsRuleSchedule{index}</li><li>**CloudWatch Event**: {normalizedFunctionName}LambdaPermissionEventsRuleCloudWatchEvent{index}</li><li>**CloudWatch Log**: {normalizedFunctionName}LambdaPermissionLogsSubscriptionFilterCloudWatchLog{index}</li><li>**IoT**: {normalizedFunctionName}LambdaPermissionIotTopicRule{i ...
+|Lambda::Permission     | <ul><li>**Schedule**: {normalizedFunctionName}LambdaPermissionEventsRuleSchedule{index}</li><li>**CloudWatch Event**: {normalizedFunctionName}LambdaPermissionEventsRuleCloudWatchEvent{index}</li><li>**CloudWatch Log**: {normalizedFunctionName}LambdaPermissionLogsSubscriptionFilterCloudWatchLog{index}</li><li>**IoT**: {normalizedFunctionName}LambdaPermissionIotTopicRule{i ...
```
It looks like the example is missing an index entry at the end

horike37 commented:
@pmuens 
Just fixed the example section!

pmuens commented:
Great! Thanks for updating @horike37 👍
